### ISSUE TITLE
Adding multiple operations loses metadata on the way

### ISSUE THREAD

**Issue by saif-ellafi** [Adventuresmith/dart-dice-parser#2]
Hi there,

another similar one to the previous, for example this roll `/r (10d10 #f<=2)kl2`. Adding the kl2 to count the lowest 2 works fine, but it loses the info about "failures" from the `#f<=2` . The counting needs to be done from the original roll, so adding it after the kl2 would already have dropped some information.

Cheers

**Comment by stevesea** [comment]
I'll have to review the code to figure out how best to handle it. But... maybe this points to the current API's clunkiness and lack of usability from an app. 

The metadata is probably silly to expose _only_ through the metadata in the RollResult or logging. 
It'll probably depend on the app / use-case for how a client wants to expose that metadata to the end user.

As in... maybe part of the API should include a way for the client to register callbacks on dice rolls or evaluation of expressions. Then at each level of the evaluation graph, the client might get an events and let it do interesting things. e.g. render a dice roll, remove dice from previous roll, etc.

**Comment by stevesea** [comment]
as to my comments on other issue, here's how the results from that roll look in my in-progress work: 

```console
❯ dart run example/main.dart -r 1234 -o pretty '(10d10 #f<=2)kl2'
(((10d10) #f<= 2) kl 2) ===> RollSummary(total: 4, results: [1, 3], metadata: {rolled: [1, 3, 3, 4, 6, 6, 7, 7, 8, 9], discarded: [3, 4, 6, 6, 7, 7, 8, 9], score: {failures: [1]}})
  (((10d10) #f<= 2) kl 2) =drop=> RollResult(total: 4, results: [1, 3], metadata: {discarded: [3, 4, 6, 6, 7, 7, 8, 9]})
      ((10d10) #f<= 2) =count=> RollResult(total: 54, results: [1, 3, 3, 4, 6, 6, 7, 7, 8, 9], metadata: {score: {failures: [1]}})
          (10d10) =rollDice=> RollResult(total: 54, results: [1, 3, 3, 4, 6, 6, 7, 7, 8, 9], metadata: {rolled: [1, 3, 3, 4, 6, 6, 7, 7, 8, 9]})
```

**Comment by stevesea** [comment]
I'm adding more documentation in the README about how to traverse the result graph.

and, I've added a way for client to register listeners. 

```dart
    // if you want to listen to every individual operation within the expression
    DiceExpression.registerListener((rollResult) {
      stdout.writeln('${rollResult.opType.name} -> $rollResult');
    });

    // if you want to listen for the RollSummary
    DiceExpression.registerSummaryListener((rollSummary) {
      stdout.writeln('$rollSummary');
    });
```

the above would register a listener for _all_ rolls.  felt like that might not be what you want -- e.g. if there's sub components of your app that do rolling outside of user interaction, felt like that'd cause lots of weirdness if you were trying to react to the events. So, rather than limiting to just the above, you can also pass the callbacks to the roll() method.

```dart 
  DiceExpression.create('2d20kh').roll(
    onRoll: (rr) => stdout.writeln('roll - $rr'),
    onSummary: (summary) => stdout.writeln('summary - $summary')
  );
```

You don't need to register the listeners unless it works for your use case... I _think_ most clients could probably just get what they need from the RollSummary.

**Comment by ELS2MU** [comment]
Hey @stevesea - that's awesome, definitely beyond what the issue was about originally but this increases the power of the library as a whole :)

As per the original report, the goal was that in the very end, the metadata of the final roll would keep the in intermediate step metadata, which is awesome to see in your example. I guess the only possible concern is if there are "multiple" formulas that get "discarded" dice key which could then overwrite or conflict. 

I haven't tried but not sure what happens if you discard highest and then discard lowest, will it add up all discarded altogether?

**Comment by saif-ellafi** [comment]
Hey @stevesea  - that's awesome, definitely beyond what the issue was about originally but this increases the power of the library as a whole :)

As per the original report, the goal was that in the very end, the metadata of the final roll would keep the in intermediate step metadata, which is awesome to see in your example. I guess the only possible concern is if there are "multiple" formulas that get "discarded" dice key which could then overwrite or conflict. 

I haven't tried but not sure what happens if you discard highest and then discard lowest, will it add up all discarded altogether?

**Comment by stevesea** [comment]
yes, each drop is its own RollREsult in the graph. But, the summary will accumulate both the drop-high and drop-low

```console
((((4d6) -h ) -l ) #s= 3) ===> RollSummary(total: 7, results: [3, 4], metadata: {rolled: [1, 3, 4, 4], discarded: [4, 1], score: {successes: [3]}})
  ((((4d6) -h ) -l ) #s= 3) =count=> RollResult(total: 7, results: [3, 4], metadata: {score: {successes: [3]}})
      (((4d6) -h ) -l ) =drop=> RollResult(total: 7, results: [3, 4], metadata: {discarded: [1]})
          ((4d6) -h ) =drop=> RollResult(total: 8, results: [1, 3, 4], metadata: {discarded: [4]})
              (4d6) =rollDice=> RollResult(total: 12, results: [1, 3, 4, 4], metadata: {rolled: [1, 3, 4, 4]})

```

my thinking was that a client could look at the summary to render the results.

e.g. from `4d6 -h -l #s=3` it rolled it rolled a 1,3,4,4. 
a 4 and a 1 were discarded, and it determined '3' was success
so you could render the rolls in a way that crossed out the discards, and highlighted success:

>  ~~1~~, **^3** , ~~4~~, 4

or, the clients could traverse the result graph to give other feedback to the user, or just have the roll listeners trigger rendering 3d dice flying across their app. 
But, I am hoping the summary itself would be usable & self-contained so clients wouldn't _have_ to mess with the graph unless they choose to

**Comment by stevesea** [comment]
fixed in v7.0.0

**Comment by saif-ellafi** [comment]
Testing, so far working really great, still for now using the toJson() to get all entries, but I like the new API.
